kairos-data/src/transaction.rs

```diff
@@ -0,0 +1,137 @@
+    let conn = pool.get().await?;
+    let res = conn
+        .interact(move |conn| {
+            let mut query = transactions::table.into_boxed::<diesel::pg::Pg>();
+
+            if let Some(sender) = filter.sender {
+                query = query.filter(transactions::public_key.eq(sender));
+            }
+            if let Some(min_timestamp) = filter.min_timestamp {
+                query = query.filter(transactions::timestamp.ge(min_timestamp));
+            }
+            if let Some(max_timestamp) = filter.max_timestamp {
+                query = query.filter(transactions::timestamp.le(max_timestamp));
+            }
+            if let Some(min_amount) = filter.min_amount {
+                query = query.filter(transactions::amount.ge(min_amount));
+            }
+            if let Some(max_amount) = filter.max_amount {
+                query = query.filter(transactions::amount.le(max_amount));
+            }
+            if let Some(recipient) = filter.recipient {
+                query = query.filter(transactions::recipient.eq(recipient));
+            }
+
+            query
+                .select(Transaction::as_select())
+                .limit(500)
+                .load::<Transaction>(conn)
+        })
+        .await??;
+    Ok(res)
+}
+
+
+pub async fn insert(
+    pool: crate::Pool,
+    kairos_trx: KairosTransaction,
+) -> Result<Transaction, crate::errors::DBError> {
+    let trx = Transaction::from(kairos_trx);
+    let conn = pool.get().await?;
+    let res = conn
+        .interact(|conn| {
+            diesel::insert_into(transactions::table)
+                .values(trx)
+                .get_result::<Transaction>(conn)
+        })
+        .await??;
+    Ok(res)
+}
+
+impl From<KairosTransaction> for Transaction {
+    fn from(tx: KairosTransaction) -> Self {
+        match tx {
+            KairosTransaction::Transfer(signed_transfer) => Transaction::from(signed_transfer),
+            KairosTransaction::Withdraw(signed_withdraw) => Transaction::from(signed_withdraw),
+            KairosTransaction::Deposit(l1_deposit) => Transaction::from(l1_deposit),
+        }
+    }
+}
+
+impl From<Signed<Transfer>> for Transaction {
+    fn from(signed_transfer: Signed<Transfer>) -> Self {
+        Transaction {
+            timestamp: Utc::now().naive_utc(),
+            public_key: hex::encode(&signed_transfer.public_key),
+            nonce: Some(signed_transfer.nonce as i64),
+            trx: TRANSFER_TRX,
+            amount: signed_transfer.transaction.amount as i64,
+            recipient: Some(hex::encode(&signed_transfer.transaction.recipient)),
+        }
+    }
+}
+
+impl From<Signed<Withdraw>> for Transaction {
+    fn from(signed_withdraw: Signed<Withdraw>) -> Self {
+        Transaction {
+            timestamp: Utc::now().naive_utc(),
+            public_key: hex::encode(&signed_withdraw.public_key),
+            nonce: Some(signed_withdraw.nonce as i64),
```

Comment:
`u64 as i64` may not preserve the value.
You could use `try_into` and error, but it's not ideal.

Try using `NUMERIC`.
https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL
https://github.com/launchbadge/sqlx/discussions/2977#discussioncomment-8428838

Reply:
NUMERIC in Diesel is represented by BigDecimal. I swapped it over to BigDecimal